b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,384 @@
+                  .createInstance(Ci.nsIConverterInputStream);
+  this._input.init(input, "UTF-8", 0,
+                   Ci.nsIConverterInputStream.DEFAULT_REPLACEMENT_CHARACTER);
+
+  this._output = Cc["@mozilla.org/intl/converter-output-stream;1"]
+                   .createInstance(Ci.nsIConverterOutputStream);
+  this._output.init(output, "UTF-8", 0, 0x0000);
+
+  // This allows a connection to disambiguate between a peer initiating a
+  // close and the socket being forced closed on shutdown.
+  this._closed = false;
+}
+Connection.prototype = {
+  // Closes this connection's input/output streams
+  close: function() {
+    if (this._closed) {
+      return;
+    }
+
+    DEBUG && debug("Close connection " + this.connectionId);
+
+    this.input.close();
+    this.output.close();
+    this._closed = true;
+
+    this.server._connectionClosed(this);
+  },
+
+  // nsIInputStreamCallback
+  onInputStreamReady: function(input) {
+    DEBUG && debug("onInputStreamReady(input=" + input + ") on thread " +
+                   Services.tm.currentThread + " (main is " +
+                   Services.tm.mainThread + ")");
+
+    try {
+      let available = 0, numChars = 0, fullMessage = "";
+
+      // Read and concat messages from input stream buffer
+      do {
+        let partialMessage = {};
+
+        available = input.available();
+        numChars = this._input.readString(available, partialMessage);
+
+        fullMessage += partialMessage.value;
+      } while(numChars < available);
+
+      if (fullMessage.length > 0) { // While readString contains something
+        // Handle concatenated JSON string
+        fullMessage = this._sanitizeMessage(fullMessage);
+        debug(fullMessage);
+
+        try {
+          // Parse JSON string to event objects
+          let events = JSON.parse(fullMessage);
+
+          events.forEach((event) => {
+            this.eventHandler.handleEvent(event);
+          });
+        } catch (e) {
+          DEBUG && debug ("Parse event error, drop this message, error: " + e.message);
+        }
+      }
+    } catch (e) {
+      if (streamClosed(e)) {
+        DEBUG && debug("WARNING: unexpected error when reading from socket; will " +
+                       "be treated as if the input stream had been closed");
+        DEBUG && debug("WARNING: actual error was: " + e.message);
+      }
+
+      // Input has been closed, but we're still expecting to read more data.
+      // available() will throw in this case, destroy the connection.
+      DEBUG && debug("onInputStreamReady called on a closed input, destroying connection, error: " + e.message);
+      this._connection.close();
+      return;
+    }
+
+    // Wait next message
+    input.asyncWait(this, 0, 0, Services.tm.currentThread);
+  },
+
+  _sanitizeMessage: function(message) {
+    // If message doesn't start with "{", discard the part before first found "}{"
+    if(!message.startsWith("{") && message.indexOf("}{") >= 0) {
```

Comment:
If `"}{"` is not found, it means the entire string is a partial JSON string that can be discard entirely. Not sure if you aware of this edge case, though it will be catch later by the try-catch while invoking `JSON.parse`.